modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+    esc2_entries = query_ldap_server(esc2_raw_filter, attributes, base_prefix: base_prefix)
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC2 vulnerable certificate templates!') if esc2_entries.blank?
+
+    # Grab a list of certificates that contain vulnerable settings.
+    # Also print out the list of SIDs that can enroll in that server.
+
+    esc2_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
+        @vuln_cert_template_table << ['ESC2', entry[:cn][0], entry[:dn][0], convert_sids_to_human_readable_name(allowed_sids)]
+      end
+      parse_dacl_or_sacl(security_descriptor.sacl) if security_descriptor.sacl
+    end
+  end
+
+  def find_esc3_vuln_cert_templates
+    esc3_template_1_raw_filter = '(&'\
+      '(objectclass=pkicertificatetemplate)'\
+      '(!(mspki-enrollment-flag:1.2.840.113556.1.4.804:=2))'\
+      '(|'\
+        '(mspki-ra-signature=0)'\
+        '(!(mspki-ra-signature=*))'\
+      ')'\
+      '(pkiextendedkeyusage=1.3.6.1.4.1.311.20.2.1)'\
+    ')'
+
+    # Find the first vulnerable types of templates, those that have the OID of the
+    # Certificate Request Agent which allows the template to be used for
+    # requesting other certificate templates on behalf of other principals.
+    attributes = ['cn', 'description', 'ntSecurityDescriptor']
+    base_prefix = 'CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration'
+    esc3_template_1_entries = query_ldap_server(esc3_template_1_raw_filter, attributes, base_prefix: base_prefix)
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC3 template 1 vulnerable certificate templates!') if esc3_template_1_entries.blank?
+
+    # Grab a list of certificates that contain vulnerable settings.
+    # Also print out the list of SIDs that can enroll in that server.
+
+    esc3_template_1_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
+        @vuln_cert_template_table << ['ESC3_TEMPLATE_1', entry[:cn][0], entry[:dn][0], convert_sids_to_human_readable_name(allowed_sids)]
+      end
```

Comment:
`flag_allowed_to_enroll` issue again. This code is super common and has been copied around into a few places.

Reply:
Fixed this up with https://github.com/rapid7/metasploit-framework/commit/fd33cdc58e47576823a874c5f17d81b3e5660f9b which uses the new function you suggested and also fixes this issue and some related ones you mentioned.